Nethermind crashes on first start with SetupKeyStore error if keystore missing

## Problem
Nethermind container crashes on first startup if keystore directory doesn't exist or is empty, with `SetupKeyStore` null reference error.

## Evidence
From xdc02 Nethermind first run:
```
[ERROR] Unhandled exception. System.NullReferenceException: Object reference not set to instance of object.
   at Nethermind.KeyStore.KeyStorePasswordProvider.SetupKeyStore() 
   at Nethermind.Runner.Program.Main(String[] args)
```

## Root Cause
- Nethermind expects keystore directory to exist
- If missing, crashes instead of creating it
- No graceful initialization like geth's `account new`

## Impact
- First-time setup fails
- Requires manual keystore creation
- Poor UX for fresh installs

## Proposed Solution
1. Create keystore directory in docker-compose:
   ```yaml
   volumes:
     - ../mainnet/nethermind-db:/nethermind/data
     - ../mainnet/.xdc-node/keystore:/nethermind/keystore  # <-- Add this
   ```

2. Pre-create in setup script:
   ```bash
   mkdir -p "$NETWORK_DIR/.xdc-node/keystore"
   ```

3. Or patch Nethermind config to disable keystore requirement for fullnode:
   ```json
   {
     "KeyStore": {
       "KeyStoreDirectory": "/nethermind/keystore",
       "PasswordFiles": [],
       "UnlockAccounts": []
     }
   }
   ```

## Files Affected
- `docker/docker-compose.nethermind.yml`
- `docker/nethermind/config.json`
- `setup.sh`

--- a/docker/erigon/start-erigon.sh
+++ b/docker/erigon/start-erigon.sh
@@ -39,6 +39,35 @@ echo "Datadir: $DATADIR"
 echo "Config: $CONFIG_FILE"
 
 # ============================================================
+# Issue #547 & #552: Pre-flight checks for config files
+# ============================================================
+echo "Pre-flight: checking config file integrity..."
+
+# Ensure bootnodes.list exists and is a file, not a directory
+if [[ -d "$BOOTNODES_FILE" ]]; then
+    echo "ERROR: $BOOTNODES_FILE is a directory (Docker mount error)"
+    echo "Removing directory and creating empty file..."
+    rm -rf "$BOOTNODES_FILE"
+    touch "$BOOTNODES_FILE"
+fi
+
+if [[ ! -f "$BOOTNODES_FILE" ]]; then
+    echo "Warning: bootnodes.list not found, creating empty file"
+    touch "$BOOTNODES_FILE"
+fi
+
+# Ensure static-nodes.json is not a directory
+STATIC_NODES="/static-nodes.json"
+if [[ -d "$STATIC_NODES" ]]; then
+    echo "ERROR: static-nodes.json is a directory (Docker mount error)"
+    echo "Removing directory and creating empty file..."
+    rm -rf "$STATIC_NODES"
+    echo "[]" > "$STATIC_NODES"
+fi
+
+echo "✓ Pre-flight checks complete"
+
+# ============================================================
 # Load config.toml - section-aware TOML parser
 # ============================================================
 load_config() {

--- a/docker/geth-pr5/start-geth-pr5.sh
+++ b/docker/geth-pr5/start-geth-pr5.sh
@@ -42,6 +42,24 @@ echo "Config: sync=$SYNC_MODE gc=$GC_MODE network=$NETWORK_ID"
 # ============================================================
 # Init genesis if needed
 # ============================================================
+
+# Issue #548: Pre-flight check for genesis.json
+if [ ! -f "$GENESIS_FILE" ]; then
+    echo "=============================================="
+    echo "ERROR: Genesis file not found!"
+    echo "=============================================="
+    echo "Expected location: $GENESIS_FILE"
+    echo ""
+    echo "This file is required to initialize the blockchain."
+    echo "Without it, the node will sync to Ethereum mainnet instead of XDC Network."
+    echo ""
+    echo "Please ensure genesis.json is mounted correctly in docker-compose.yml:"
+    echo "  volumes:"
+    echo "    - ./mainnet/genesis.json:/work/genesis.json:ro"
+    echo "=============================================="
+    exit 1
+fi
+
 if [ ! -d "$DATADIR/XDC/chaindata" ]; then
     echo "No existing chaindata found, initializing..."
     
@@ -54,13 +72,56 @@ if [ ! -d "$DATADIR/XDC/chaindata" ]; then
     fi
     
     # Initialize genesis
-    if [ -f "$GENESIS_FILE" ]; then
-        echo "Initializing Genesis Block..."
-        XDC init --datadir "$DATADIR" "$GENESIS_FILE"
-        echo "Genesis initialized successfully"
+    echo "Initializing Genesis Block..."
+    XDC init --datadir "$DATADIR" "$GENESIS_FILE"
+    echo "Genesis initialized successfully"
+    
+    # Issue #548: Verify genesis hash after initialization
+    echo "Verifying genesis block hash..."
+    sleep 2
+    
+    # Query block 0 to get genesis hash
+    # Start XDC in background temporarily to query genesis
+    XDC --datadir "$DATADIR" --networkid ${NETWORK_ID:-50} --port 0 --nodiscover --maxpeers 0 \
+        --http --http.addr 127.0.0.1 --http.port 18545 --http.api eth &
+    XDC_PID=$!
+    
+    # Wait for RPC to become available
+    for i in $(seq 1 10); do
+        if curl -sf -X POST http://127.0.0.1:18545 \
+            -H "Content-Type: application/json" \
+            -d '{"jsonrpc":"2.0","method":"net_version","params":[],"id":1}' >/dev/null 2>&1; then
+            break
+        fi
+        sleep 1
+    done
+    
+    # Get genesis block hash
+    GENESIS_HASH=$(curl -sf -X POST http://127.0.0.1:18545 \
+        -H "Content-Type: application/json" \
+        -d '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x0",false],"id":1}' 2>/dev/null \
+        | grep -o '"hash":"0x[^"]*"' | cut -d'"' -f4 | sed 's/^0x//')
+    
+    # Stop temporary XDC instance
+    kill $XDC_PID 2>/dev/null || true
+    wait $XDC_PID 2>/dev/null || true
+    
+    # Expected XDC mainnet genesis hash
+    EXPECTED_GENESIS="4a9d748bd78a8d0385b67788c2435dcdb914f98a96250b68863a1f8b7642d6b1"
+    
+    if [ -n "$GENESIS_HASH" ]; then
+        echo "Genesis hash: 0x$GENESIS_HASH"
+        if [ "$GENESIS_HASH" = "$EXPECTED_GENESIS" ]; then
+            echo "✓ Genesis hash verified (XDC mainnet)"
+        else
+            echo "⚠ Warning: Genesis hash does not match XDC mainnet"
+            echo "  Got:      0x$GENESIS_HASH"
+            echo "  Expected: 0x$EXPECTED_GENESIS"
+            echo "  Proceeding anyway (may be testnet/devnet)..."
+        fi
     else
-        echo "ERROR: No genesis file at $GENESIS_FILE"
-        exit 1
+        echo "⚠ Warning: Could not verify genesis hash (RPC unavailable)"
+        echo "  Node will start but please verify network after sync"
     fi
 else
     echo "Existing chaindata found at $DATADIR/XDC/chaindata"

--- a/docker/nethermind/start-nethermind.sh
+++ b/docker/nethermind/start-nethermind.sh
@@ -51,6 +51,60 @@ if [ ! -f "$DATADIR/.node-identity" ]; then
   echo "[SkyNet] Generated identity seed (coinbase will be read from RPC after start)"
 fi
 
+# Issue #557: Ensure KZG trusted setup file exists
+# Nethermind requires kzg_trusted_setup.txt for EIP-4844 (blob transactions)
+KZG_FILE="/nethermind/kzg_trusted_setup.txt"
+if [[ ! -f "$KZG_FILE" ]]; then
+    echo "⚠ Warning: kzg_trusted_setup.txt not found at $KZG_FILE"
+    echo "Attempting to copy from embedded resources..."
+    
+    # Try to copy from Nethermind binary resources
+    if [[ -f "/nethermind/Data/kzg_trusted_setup.txt" ]]; then
+        cp "/nethermind/Data/kzg_trusted_setup.txt" "$KZG_FILE"
+        echo "✓ Copied KZG setup from /nethermind/Data/"
+    elif [[ -f "/nethermind/data/kzg_trusted_setup.txt" ]]; then
+        cp "/nethermind/data/kzg_trusted_setup.txt" "$KZG_FILE"
+        echo "✓ Copied KZG setup from /nethermind/data/"
+    else
+        # Download from official Ethereum KZG ceremony
+        echo "Downloading KZG trusted setup from Ethereum Foundation..."
+        if command -v curl >/dev/null 2>&1; then
+            curl -fsSL -o "$KZG_FILE" \
+                "https://raw.githubusercontent.com/ethereum/c-kzg-4844/main/src/trusted_setup.txt" 2>/dev/null || \
+            curl -fsSL -o "$KZG_FILE" \
+                "https://github.com/ethereum/consensus-specs/raw/dev/presets/mainnet/trusted_setups/trusted_setup.txt" 2>/dev/null || \
+            echo "ERROR: Failed to download KZG trusted setup"
+        elif command -v wget >/dev/null 2>&1; then
+            wget -q -O "$KZG_FILE" \
+                "https://raw.githubusercontent.com/ethereum/c-kzg-4844/main/src/trusted_setup.txt" 2>/dev/null || \
+            wget -q -O "$KZG_FILE" \
+                "https://github.com/ethereum/consensus-specs/raw/dev/presets/mainnet/trusted_setups/trusted_setup.txt" 2>/dev/null || \
+            echo "ERROR: Failed to download KZG trusted setup"
+        fi
+        
+        if [[ -f "$KZG_FILE" && -s "$KZG_FILE" ]]; then
+            echo "✓ Downloaded KZG trusted setup ($(wc -c < "$KZG_FILE") bytes)"
+        else
+            echo "ERROR: KZG trusted setup file is missing or empty"
+            echo "Nethermind may crash with 'SetupKeyStore' or 'KZG' errors"
+            echo ""
+            echo "Manual fix: Copy kzg_trusted_setup.txt to docker volume:"
+            echo "  docker cp kzg_trusted_setup.txt <container>:/nethermind/kzg_trusted_setup.txt"
+            # Don't exit - let Nethermind try to start anyway (may work on older versions)
+        fi
+    fi
+else
+    echo "✓ KZG trusted setup file exists ($(wc -c < "$KZG_FILE") bytes)"
+fi
+
+# Issue #557: Ensure keystore directory exists
+KEYSTORE_DIR="/nethermind/keystore"
+if [[ ! -d "$KEYSTORE_DIR" ]]; then
+    echo "Creating keystore directory: $KEYSTORE_DIR"
+    mkdir -p "$KEYSTORE_DIR"
+    chmod 700 "$KEYSTORE_DIR"
+fi
+
 # Check if chainspec exists
 if [[ ! -f /nethermind/chainspec/xdc.json ]]; then
     echo "ERROR: Chainspec file not found at /nethermind/chainspec/xdc.json"

--- a/setup.sh
+++ b/setup.sh
@@ -71,15 +71,19 @@ init_logging() {
     chmod 640 "$LOG_FILE" 2>/dev/null || true
 }
 
-# Ensure path is a file, not a directory (Docker creates dirs for missing volume mounts)
+# Issue #547 & #552: Ensure path is a file, not a directory
+# Docker creates directories for missing volume mount sources
 ensure_file_path() {
     local f="$1"
     if [[ -d "$f" ]]; then
+        echo "  Removing directory: $f (Docker will mount this as file)"
         rm -rf "$f"
     fi
     mkdir -p "$(dirname "$f")"
     # Actually create the file to prevent Docker from creating it as a directory
-    touch "$f"
+    if [[ ! -f "$f" ]]; then
+        touch "$f"
+    fi
 }
 
 log() {
@@ -1577,22 +1581,84 @@ start_services() {
         fi
     done
     
-    # Verify critical files exist and are actual files (not directories)
-    for f in mainnet/start-node.sh mainnet/genesis.json mainnet/.pwd; do
+    # Issue #547 & #552: Verify critical files exist and are actual files (not directories)
+    # Docker creates directories when volume mount source is missing
+    log "Pre-flight check: ensuring config files are files, not directories..."
+    
+    # Network-specific files to check
+    local config_files=(
+        "${NETWORK}/start-node.sh"
+        "${NETWORK}/genesis.json"
+        "${NETWORK}/.pwd"
+        "${NETWORK}/bootnodes.list"
+    )
+    
+    # Additional config files that Docker might create as directories
+    local state_config_files=(
+        "../${NETWORK}/.xdc-node/skynet.conf"
+        "../${NETWORK}/.xdc-node/config.toml"
+        "../${NETWORK}/.xdc-node/.env"
+    )
+    
+    for f in "${config_files[@]}"; do
         local fpath="$PROJECT_ROOT/docker/$f"
         if [[ -d "$fpath" ]]; then
             warn "$f was created as a directory (Docker artifact). Removing and recreating..."
             rm -rf "$fpath"
         fi
-        if [[ ! -f "$fpath" || ! -s "$fpath" ]]; then
-            warn "$f is missing or empty. Re-downloading..."
-            local base_url="https://raw.githubusercontent.com/XinFinOrg/XinFin-Node/master/mainnet"
-            local fname=$(basename "$f")
-            curl -fsSL "$base_url/$fname" -o "$fpath" 2>/dev/null || warn "Failed to download $fname"
-            [[ "$fname" == "start-node.sh" ]] && chmod +x "$fpath" 2>/dev/null || true
-            [[ "$fname" == ".pwd" ]] && { touch "$fpath"; chmod 600 "$fpath"; }
+        
+        # Ensure parent directory exists
+        mkdir -p "$(dirname "$fpath")"
+        
+        if [[ ! -f "$fpath" ]]; then
+            case "$(basename "$f")" in
+                start-node.sh|genesis.json|bootnodes.list)
+                    warn "$f is missing. Attempting to download..."
+                    local base_url="https://raw.githubusercontent.com/XinFinOrg/XinFin-Node/master/${NETWORK}"
+                    local fname=$(basename "$f")
+                    if ! curl -fsSL --connect-timeout 10 "$base_url/$fname" -o "$fpath" 2>/dev/null || [[ ! -s "$fpath" ]]; then
+                        # Fallback: try mainnet if network-specific file not found
+                        [[ "$NETWORK" != "mainnet" ]] && curl -fsSL --connect-timeout 10 \
+                            "https://raw.githubusercontent.com/XinFinOrg/XinFin-Node/master/mainnet/$fname" \
+                            -o "$fpath" 2>/dev/null || true
+                    fi
+                    [[ "$fname" == "start-node.sh" ]] && chmod +x "$fpath" 2>/dev/null || true
+                    ;;
+                .pwd)
+                    touch "$fpath"
+                    chmod 600 "$fpath"
+                    log "Created empty password file: $f"
+                    ;;
+            esac
+        fi
+        
+        # Final check: if still directory or doesn't exist, create empty file
+        if [[ -d "$fpath" ]]; then
+            rm -rf "$fpath"
+            touch "$fpath"
+            warn "Force-created $f as empty file (was directory)"
+        elif [[ ! -f "$fpath" ]]; then
+            touch "$fpath"
+            info "Created placeholder for: $f"
         fi
     done
+    
+    # Check state config files (relative to docker dir)
+    for f in "${state_config_files[@]}"; do
+        local fpath="$PROJECT_ROOT/docker/$f"
+        if [[ -d "$fpath" ]]; then
+            warn "$(basename "$fpath") was created as directory. Removing..."
+            rm -rf "$fpath"
+        fi
+        # These are created by other setup functions, just ensure they're not directories
+        if [[ ! -f "$fpath" ]]; then
+            mkdir -p "$(dirname "$fpath")"
+            touch "$fpath"
+            info "Pre-created: $(basename "$fpath")"
+        fi
+    done
+    
+    log "✓ Config file pre-flight check complete"
     
     # Also remove any Docker-created directories in the data volume
     local container_name="xdc-node"

